data/transactions/currency.go

```diff
@@ -0,0 +1,360 @@
+		return basics.CurrencyParams{}, fmt.Errorf("currency index %d not found in account %v", cid.Index, cid.Creator)
+	}
+
+	return curr, nil
+}
+
+func (cc CurrencyConfigTxnFields) apply(header Header, balances Balances, spec SpecialAddresses, ad *ApplyData, txnCounter uint64) error {
+	if cc.ConfigCurrency == (basics.CurrencyID{}) {
+		// Allocating a currency.
+		record, err := balances.Get(header.Sender)
+		if err != nil {
+			return err
+		}
+		record.Currencies = clone(record.Currencies)
+		record.CurrencyParams = cloneParams(record.CurrencyParams)
+
+		// Ensure index is never zero
+		newidx := txnCounter + 1
+
+		// Sanity check that there isn't a currency with this counter value.
+		_, present := record.CurrencyParams[newidx]
+		if present {
+			return fmt.Errorf("already found a currency with index %d", newidx)
+		}
+
+		cid := basics.CurrencyID{
+			Creator: header.Sender,
+			Index:   newidx,
+		}
+
+		record.CurrencyParams[cid.Index] = cc.CurrencyParams
+		record.Currencies[cid] = basics.CurrencyHolding{
+			Amount: cc.CurrencyParams.Total,
+		}
+
+		if len(record.Currencies) > balances.ConsensusParams().MaxCurrenciesPerAccount {
+			return fmt.Errorf("too many currencies in account: %d > %d", len(record.Currencies), balances.ConsensusParams().MaxCurrenciesPerAccount)
+		}
+
+		return balances.Put(record)
+	}
+
+	// Re-configuration and destroying must be done by the manager key.
+	params, err := getParams(balances, cc.ConfigCurrency)
+	if err != nil {
+		return err
+	}
+
+	if header.Sender != params.Manager {
+		return fmt.Errorf("transaction issued by %v, not manager key %v", header.Sender, params.Manager)
+	}
+
+	record, err := balances.Get(cc.ConfigCurrency.Creator)
+	if err != nil {
+		return err
+	}
+
+	record.Currencies = clone(record.Currencies)
+	record.CurrencyParams = cloneParams(record.CurrencyParams)
+
+	if cc.CurrencyParams == (basics.CurrencyParams{}) {
+		// Destroying a currency.  The creator account must hold
+		// the entire outstanding currency amount.
+		if record.Currencies[cc.ConfigCurrency].Amount != params.Total {
+			return fmt.Errorf("cannot destroy currency: holding only %d/%d", record.Currencies[cc.ConfigCurrency].Amount, params.Total)
+		}
+
+		delete(record.Currencies, cc.ConfigCurrency)
+		delete(record.CurrencyParams, cc.ConfigCurrency.Index)
+	} else {
+		// Changing keys in a currency.
+		if !params.Manager.IsZero() {
+			params.Manager = cc.CurrencyParams.Manager
+		}
+		if !params.Reserve.IsZero() {
+			params.Reserve = cc.CurrencyParams.Reserve
+		}
+		if !params.Freeze.IsZero() {
+			params.Freeze = cc.CurrencyParams.Freeze
+		}
+		if !params.Clawback.IsZero() {
+			params.Clawback = cc.CurrencyParams.Clawback
+		}
+
+		record.CurrencyParams[cc.ConfigCurrency.Index] = params
+	}
+
+	return balances.Put(record)
+}
+
+func (ct CurrencyTransferTxnFields) apply(header Header, balances Balances, spec SpecialAddresses, ad *ApplyData) error {
+	// Default to sending from the transaction sender's account.
```

Comment:
Maybe break this function into smaller pieces of code?
It took me about 45 minutes to read through it and figure out all the cases it covers.

Reply:
I tried to refactor it a bit (see pushed code).